src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java

```diff
@@ -255,6 +255,11 @@ private CacheResult analyzeResult(CacheContext context, List<MojoExecution> mojo
                         build.getCacheImplementationVersion());
             }
 
+            if (lifecyclePhasesHelper.isLaterPhaseThanBuild("package", build)) {
+                LOGGER.warn("Cached build doesn't include phase 'package', cannot restore");
```

Comment:
The expected behavior is that when two consecutive 'mvn compile' commands are used, the second 'mvn compile' should use the cache. I'm concerned that this implementation might lead to a performance regression. A more accurate implementation would be to always attach the generated sources in `save` and process partial success more accurately.

In the current implementation, the 'analyzeResult' returns partial success here (see lines 277-278) as expected. The fix should be made in the partial success processing logic. A build without a package phase should restore generated sources only and rerun all the missing lifecycle phases to build artifacts.